test/unit/org/apache/cassandra/metrics/ClientRequestSizeMetricsTest.java

```diff
@@ -0,0 +1,127 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.metrics;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Snapshot;
+
+import org.apache.cassandra.cql3.CQLTester;
+import org.apache.cassandra.metrics.ClearableHistogram;
+import org.apache.cassandra.metrics.ClientRequestSizeMetrics;
+import org.apache.cassandra.metrics.DecayingEstimatedHistogramReservoir.EstimatedHistogramReservoirSnapshot;
+import org.apache.cassandra.metrics.DecayingEstimatedHistogramReservoir.Range;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Ensures we properly account for metrics tracked in the native protocol
+ */
+public class ClientRequestSizeMetricsTest extends CQLTester
+{
+    @BeforeClass
+    public static void setUp()
+    {
+        requireNetwork();
+    }
+
+    @Test
+    public void testReadAndWriteMetricsAreRecordedDuringNativeRequests() throws Throwable
+    {
+        // We want to ignore all the messages sent by the driver upon connection as well as
+        // the event sent upon schema updates
+        clearMetrics();
+
+        executeNet("SELECT * from system.peers");
+        assertThat(ClientRequestSizeMetrics.totalBytesRead.getCount()).isGreaterThan(0);
+        assertThat(ClientRequestSizeMetrics.totalBytesWritten.getCount()).isGreaterThan(0);
+
+        // The request fit in one single frame so we know the new number of received frames
+        assertEquals(1, ClientRequestSizeMetrics.bytesReceivedPerFrame.getCount());
+
+        long requestLength = ClientRequestSizeMetrics.totalBytesRead.getCount();
+        long responseLength = ClientRequestSizeMetrics.totalBytesWritten.getCount();
+
+        Snapshot snapshot = ClientRequestSizeMetrics.bytesReceivedPerFrame.getSnapshot();
+        assertMin(snapshot, requestLength);
+        assertMax(snapshot, requestLength);
+
+        // The response fit in one single frame so we know the new number of received frames
+        assertEquals(1, ClientRequestSizeMetrics.bytesTransmittedPerFrame.getCount());
+
+        snapshot = ClientRequestSizeMetrics.bytesTransmittedPerFrame.getSnapshot();
+        assertMin(snapshot, responseLength);
+        assertMax(snapshot, responseLength);
+
+        // Let's fire the same request again and test that the changes are the same that previously
```

Comment:
Can't you avoid duplicating code by folding into a loop such as:
```
for (int i=1; i <=2 ; i++)
  {
    long expectedTotalBytesRead = i * requestLength;
    ....
  }
```

Reply:
There is some differences between the 2 set of checks so it is not really feasable to use a loop but I can extract a method to reduce the code duplication.